### PR TITLE
Move globbing logic out of TChain::Add

### PR DESCRIPTION
**Pull request by vepadulano**
First attempt at moving the globbing logic in `TChain::Add` out of the method. The main remaining question is where will this new free function actually live. It depends on `TRegexp`, `TString` and `TSystem`, so I guess  somewhere in `core/base`. Any ideas here?

I removed usage of `TList` in the function, with C++17 support I could use `std::filesystem` instead of `gSystem` which could be another small improvement in the logic.

### CODE REVIEW

**Comment by eguiraud** [comment]
I think it should live in InternalTreeUtils?

**Comment by eguiraud** [comment]
How can we be sure that this function returns the same files _in the same order_ as the previous one in all cases?

**Comment by eguiraud** [comment]
I only took a quick look but I see at least one major difference between the old and the new implementation: the new one throws if a directory cannot be opened. I agree that in principle that's better, but it's a backward-incompatible change.

**Comment by vepadulano** [comment]
> I think it should live in InternalTreeUtils?

I was thinking this is generic enough that we may want to use it in places that do not strictly depend from `libTree`, but maybe I'm just overthinking.

> How can we be sure that this function returns the same files in the same order as the previous one in all cases?

I hope we already get a (very good?) degree of certainty by not breaking existing test cases. I can come up with a few more just to add extra safety. Irrespective of this, the function is taken verbatim from the logic in `TChain::Add`. Do you see any part of the function that intuitively might lead to different results? I was thinking about this and the only place that remotely gives me a slight doubt is using `std::sort` instead of `TList::Sort`, but I really hope that doesn't introduce a different behaviour

>  the new one throws if a directory cannot be opened.

Yes indeed that's a different behaviour. The existing logic never throws. In case `dir == nullptr`, which I guess happens if the expanded directory cannot be found/opened (?), then it will just fall to the end of the `TChain::Add` method and `return nf`, where `nf` will be zero since no files have been added. In this regard I see two options:

1. The new function throws (because subjectively is the correct behaviour). Then in `TChain::Add` we can keep the old behaviour, by catching the error and not re-raising it.
2. The new function can return an empty vector, in case `dir == nullptr`. This is more similar in spirit with the old behaviour.

**Comment by eguiraud** [comment]
> I was thinking this is generic enough that we may want to use it in places that do not strictly depend from libTree

I guess if the need ever comes up we can (carefully!) consider making this particular globbing implementation available outside of `tree/`, but I'd be very conservative. How does RNTuple do globbing?

> I hope we already get a (very good?) degree of certainty by not breaking existing test cases.

Do we have tests for edge cases in globbing or rarely-used features of the globbing syntax?

> the only place that remotely gives me a slight doubt is using std::sort instead of TList::Sort

I would keep `TList::Sort` to avoid unintended shuffling of things with weird characters or stuff like that. Otherwise you'd have to make sure that stuff like `01.root`, `001.root` and `0.root` get ordered the same way between the two. But why risking that?

**Comment by vepadulano** [comment]
@eguiraud thanks for the inputs. I modified the previous commit and now the PR is ready for review. More tests were added (including one for remote globbing that we didn't test before). I will add documentation to the function once we agree on the rest

File: tree/tree/inc/ROOT/InternalTreeUtils.hxx [hunk 1]
```diff
@@ -78,6 +78,8 @@ public:
 std::unique_ptr<TChain> MakeChainForMT(const std::string &name = "", const std::string &title = "");
 std::vector<std::unique_ptr<TChain>> MakeFriends(const ROOT::TreeUtils::RFriendInfo &finfo);
 
+std::vector<std::string> ExpandGlob(TString glob);
```
Comment (eguiraud) [inline tree/tree/inc/ROOT/InternalTreeUtils.hxx:81]: why a TString and why by copy? :smile:

File: tree/tree/src/InternalTreeUtils.cxx [hunk 2]
```diff
@@ -403,6 +408,56 @@ std::vector<std::unique_ptr<TChain>> MakeFriends(const ROOT::TreeUtils::RFriendI
    return friends;
 }
 
+std::vector<std::string> ExpandGlob(TString glob)
```
Comment (eguiraud) [inline tree/tree/src/InternalTreeUtils.cxx:411]: needs docs

File: tree/tree/src/TChain.cxx [hunk 3]
```diff
@@ -412,13 +385,18 @@ Int_t TChain::Add(const char* name, Long64_t nentries /* = TTree::kMaxEntries */
...
+            nf += AddFile(path.c_str(), nentries, treename);
          } else {
-            nf += AddFile(TString::Format("%s/%s%s", directory.Data(), file, suffix.Data()), nentries);
+            nf += AddFile(TString::Format("%s%s", path.c_str(), suffix.Data()), nentries);
          }
       }
-      l.Delete();
+   } catch (const std::runtime_error &) {
+      // The 'ExpandGlob' function may throw in case the directory from the glob
+      // cannot be opened. We keep the old behaviour in this case, i.e. we
```
Comment (eguiraud) [inline tree/tree/src/TChain.cxx:395]: readers won't know what you mean with "the old behavior"

File: tree/tree/src/TChain.cxx [hunk 4]
```diff
@@ -371,38 +372,10 @@ Int_t TChain::Add(const char* name, Long64_t nentries /* = TTree::kMaxEntries */
...
-         TString s = file;
-         if ( (basename!=file) && s.Index(re) == kNPOS) continue;
-         l.Add(new TObjString(file));
-      }
-      gSystem->FreeDirectory(dir);
-      //sort the files in alphanumeric order
-      l.Sort();
-      TIter next(&l);
-      TObjString *obj;
+   try {
```
Comment (eguiraud) [inline tree/tree/src/TChain.cxx:375]: the `try` seems to encompass too much code, can it include just the `ExpandGlob` call?

File: tree/tree/src/InternalTreeUtils.cxx [hunk 5]
```diff
@@ -403,6 +408,56 @@ std::vector<std::unique_ptr<TChain>> MakeFriends(const ROOT::TreeUtils::RFriendI
...
+      TString s;
+      while ((file = gSystem->GetDirEntry(dir))) {
+         if (!strcmp(file, ".") || !strcmp(file, ".."))
+            continue;
+         s = file;
+         if ((glob != file) && s.Index(re) == kNPOS)
+            continue;
+         // Using "/" as separator here as it was done in TChain::Add
+         // In principle this should be using the appropriate platform separator
+         l.Add(new TObjString((directory + "/" + file).Data()));
```
Comment (eguiraud) [inline tree/tree/src/InternalTreeUtils.cxx:443]:
```suggestion
         l.Add(new TObjString((directory + '/' + file).Data()));
```

Comment (eguiraud) [inline tree/tree/src/InternalTreeUtils.cxx:443]: I think this is leaking the TObjStrings. You have to explicitly set `l.SetOwner()`

File: tree/tree/src/InternalTreeUtils.cxx [hunk 6]
```diff
@@ -403,6 +408,56 @@ std::vector<std::unique_ptr<TChain>> MakeFriends(const ROOT::TreeUtils::RFriendI
...
+
+      // Convert TList<TString> to std::vector<std::string>
+      std::vector<std::string> ret;
+      ret.reserve(l.GetEntries());
+      for (const auto *tobjstr : ROOT::RangeStaticCast<const TObjString *>(l)) {
+         ret.push_back(tobjstr->GetName());
+      }
+      return ret;
+   } else {
+      throw std::runtime_error("Could not open directory '" + directory + "'.");
```
Comment (eguiraud) [inline tree/tree/src/InternalTreeUtils.cxx:457]: I know we don't do it consistently, but it's useful to put the context of the error in the exception message, saves some work to the person that will have to debug this:

```suggestion
      throw std::runtime_error("ExpandGlob: could not open directory '" + directory + "'.");
```

File: tree/tree/src/InternalTreeUtils.cxx [hunk 7]
```diff
@@ -403,6 +408,56 @@ std::vector<std::unique_ptr<TChain>> MakeFriends(const ROOT::TreeUtils::RFriendI
...
+   const char *epath = gSystem->ExpandPathName(directory.Data());
+   void *dir = gSystem->OpenDirectory(epath);
+   delete[] epath;
+
+   if (dir) {
+      // create a TList to store the file names (not yet sorted)
+      TList l;
+      TRegexp re(glob, true);
+      const char *file;
+      TString s;
```
Comment (eguiraud) [inline tree/tree/src/InternalTreeUtils.cxx:434]:
```suggestion
      TString fname;
```

File: tree/tree/test/TChainParsing.cxx [hunk 8]
```diff
@@ -153,3 +164,49 @@ TEST(TChainParsing, GlobbingWithTreenameToken)
...
+   const auto *chainFiles = c.GetListOfFiles();
+   ASSERT_TRUE(chainFiles);
+   EXPECT_EQ(chainFiles->GetEntries(), 0);
+
+   // Check that the equivalent call to the ExpandGlob function throws
+   try {
+      const auto expanded_glob = ROOT::Internal::TreeUtils::ExpandGlob("nonexistingpath/nonexistingfile*");
+   } catch (const std::runtime_error &err) {
+      std::string msg{"Could not open directory 'nonexistingpath'."};
+      EXPECT_STREQ(msg.c_str(), err.what());
```
Comment (eguiraud) [inline tree/tree/test/TChainParsing.cxx:183]: I think this can be simplified to:
```suggestion
      EXPECT_EQ(msg, err.what());
```

**Review by eguiraud — CHANGES_REQUESTED** [review]
Thank you Vincenzo, I think there is a memleak, other than that looks good.

File: tree/tree/inc/ROOT/InternalTreeUtils.hxx (hunk 1 above)
Comment (vepadulano) [reply tree/tree/inc/ROOT/InternalTreeUtils.hxx:81]: My reasoning was:
1. Pass by value because anyway we need a copy of the basename since we are modifying it in the function (see `glob.Remove(0, slashpos + 1);`
2. TString because it might give us the chance to also provide a modern `ExpandGlob` that uses only standard C++ functionality (probably except for `TRegexp` and we could keep the same name of the function by e.g. providing an overload that takes `std::string_view`

File: tree/tree/src/InternalTreeUtils.cxx (hunk 5 above)
Comment (vepadulano) [reply tree/tree/src/InternalTreeUtils.cxx:443]: Good catch, I just copied it over from the previous function so probably there was a leak, unless `l.Delete()` also took care of deleting the strings irrespective of the fact there was no call to `SetOwner` before

Comment (pcanal) [reply tree/tree/src/InternalTreeUtils.cxx:443]: >  unless l.Delete() also took care of deleting the strings irrespective of the fact there was no call to SetOwner before

That is correct.  `Delete` will delete.  `Clear` will delete only if it is the owner.  And `~TList` essentially calls `Clear`

File: tree/tree/src/InternalTreeUtils.cxx [hunk 9]
```diff
@@ -403,6 +408,56 @@ std::vector<std::unique_ptr<TChain>> MakeFriends(const ROOT::TreeUtils::RFriendI
...
+         // Using "/" as separator here as it was done in TChain::Add
+         // In principle this should be using the appropriate platform separator
+         l.Add(new TObjString((directory + "/" + file).Data()));
+      }
+      gSystem->FreeDirectory(dir);
+      // sort the files in alphanumeric order
+      l.Sort();
+
+      // Convert TList<TString> to std::vector<std::string>
+      std::vector<std::string> ret;
```
Comment (pcanal) [inline tree/tree/src/InternalTreeUtils.cxx:450]: The disadvantage of this interface (creating `TObjString` then copying to `std::string` rather than just returning the the `TList`) is that this increases the memory churns (where the gain is 'solely' a more modern looking interface).

File: tree/tree/inc/ROOT/InternalTreeUtils.hxx (hunk 1 above)
Comment (eguiraud) [reply tree/tree/inc/ROOT/InternalTreeUtils.hxx:81]: I see! It might be worth a comment, because this signature will make every dev that encounters it raise an eyebrow.

File: tree/tree/src/InternalTreeUtils.cxx (hunk 9 above)
Comment (vepadulano) [reply tree/tree/src/InternalTreeUtils.cxx:450]: Yeah, I am also undecided. If I keep the `TList`, then I need to document that and the caller will need to do things a bit more clunkily:
```cpp
const auto files = ExpandGlob(files);
for (const auto *objstr: ROOT::RangeStaticCast<const TObjString *>(files)){
    std::cout << objstr->GetName() << std::endl;
}
```
So I see that as a greater disadvantage at this point, but I'm a bit on the fence about it

Comment (eguiraud) [reply tree/tree/src/InternalTreeUtils.cxx:450]: I'm not sure what Philippe refers to exactly with "memory churn", the conversion will not take more than a millisecond and `ExpandGlob` is not called in a hot loop, so I'm happy to take the "hit" for a more usable interface.

Comment (pcanal) [reply tree/tree/src/InternalTreeUtils.cxx:450]: memory churns:  Each element of the list is created once with 2 allocations (TObjString and its TString)) and then the copy in the std::vector means that this is copied once more into the std::string.  Indeed, "not too serious" but still ...

> . If I keep the TList, then I need to document that and the caller will need to do things a bit more clunkily:

Actually you can solves "that" problem by using:
```
std::unique_ptr<TList*>  ExpandGlob(TString glob);  // and call SetOwner(kTRUE) inside.
```

Comment (eguiraud) [reply tree/tree/src/InternalTreeUtils.cxx:450]: the extra copy for 10k(!) non-small strings takes 0.8 milliseconds on my laptop

Comment (pcanal) [reply tree/tree/src/InternalTreeUtils.cxx:450]: In that case why would you worry about the copy of the argument? :) :)

More seriously, that still leave the interface to be 'inconsistent' (taking a `TString` and returning `std::string's).

If the copy of the return value does not cost anything, we may as well copy the argument too:
```
std::vector<std::string> ExpandGlob(std::string glob);
```

Comment (vepadulano) [reply tree/tree/src/InternalTreeUtils.cxx:450]: In a previous comment I was saying that using `TString` as input parameter has the "added benefit" that I can then create in the future another overload of this function that takes `std::string` but internally uses only C++ standard library classes (except maybe just for `TRegexp` since there is no better standard alternative).

But, we could just simply agree that this function is only for `TChain` and classes that want to replicate that behaviour. In the future, i.e. for RNTuple and ROOT 7, we can still move to another function with more modern implementation. 

It's an internal function anyway, and I believe it should stay that way, so we will be able to change name or overload it as needed when the need comes.

Comment (pcanal) [reply tree/tree/src/InternalTreeUtils.cxx:450]: I am thinking that for this PR to reduce code churn (i.e. use the code of the function 'as-is') and still have a consistent interface we should use either:
```
std::unique_ptr<TList*>  ExpandGlob(TString glob);
```
or
```
std::vector<std::string> ExpandGlob(std::string glob);  // First line of the function is TString glob = in_glob;
```

Comment (vepadulano) [reply tree/tree/src/InternalTreeUtils.cxx:470]: In the latest status of the commit I changed to a less unexpected function signature, which shouldn't need specific extra comments. Let me know what you think

**Review by pcanal — APPROVED** [review]

**Review by eguiraud — APPROVED** [review]
Thanks @vepadulano !
